File: src/22.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE MultiWayIf #-}
--{-# LANGUAGE GHC2021 #-}
--{-# LANGUAGE TypeAbstractions #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Main where

import Debug.Trace

import Data.Maybe
import Data.List
import Data.Bits

--import Data.Array (Array)
--import qualified Data.Array as A
import Control.Monad
import Control.Monad.ST
import Data.STRef
import Data.Array.ST
import Data.Array.Unboxed (Array, UArray)
--import Data.Array.MArray (modifyArray)
import qualified Data.Array.Unboxed as A

--import Data.Map.Strict (Map)
--import qualified Data.Map.Strict as M
import Data.HashMap.Lazy (HashMap)
import qualified Data.IntMap as M
import qualified Data.ByteString.Char8 as C

import Advent
import Advent.Input

main :: IO ()
main = do
    input <- readInputLines
    let
        inputNums = map (fst . fromJust . C.readInt) input
        secretSeqs = map (evolveSecret 2001) inputNums
        priceSeqs = map (\seq -> map price $ A.elems seq) secretSeqs
        changeSeqs = map (\seq -> (sequence4s (changes seq))) priceSeqs -- :: [[(Int, Int, Int, Int)]]
        totals = makeTotals2 (map (take 1997) changeSeqs) (map (drop 4) priceSeqs) -- :: HashMap (Int, Int, Int, Int) Int
    print $ sum $ map (\seq -> seq A.! 2000) $ secretSeqs
    print $ maximum $ A.elems totals

--evolveSecret :: Int -> [Int]
--evolveSecret = iterate Main.step
evolveSecret :: Int -> Int -> UArray Int Int
evolveSecret num start = runSTUArray $ do
    result <- newArray (0, num-1) 0
    x <- newSTRef start
    forM_ [0..num-1] $ \i -> do
        xv <- readSTRef x
        writeArray result i xv
        writeSTRef x $ step xv
    return result

step :: Int -> Int
step x = let
        x1 = prune $ mix x $ x `shiftL` 6
        x2 = prune $ mix x1 $ x1 `shiftR` 5
        x3 = prune $ mix x2 $ x2 `shiftL` 11
    in x3

mix x y = x `xor` y

prune x = x .&. 16777215

price x = x `mod` 10

changes :: [Int] -> [Int]
changes xs = map (uncurry $ flip (-)) $ zip xs $ tail xs

bounds = ((-9,-9,-9,-9),(9,9,9,9))
type Change = (Int, Int, Int, Int)

sequence4s :: [Int] -> [Change]
--sequence4s seq = map (\(a, b, c, d) -> (a+10)*8000 + (b+10)*400 + (c+10)*20 + (d+10)) $ zip4 (tails seq !! 3) (tails seq !! 2) (tails seq !! 1) seq
sequence4s seq = zip4 (tails seq !! 3) (tails seq !! 2) (tails seq !! 1) seq

makeSeqToPrice = map (\(prices, changes) -> M.fromListWith (\a b -> b) $ zip (take 1997 changes) (drop 4 prices))

makeTotals2 :: [[Change]] -> [[Int]] -> UArray Change Int
makeTotals2 changeSeqs priceSeqs = runSTUArray $ do
    totals :: STUArray s Change Int
        <- newArray bounds 0
    forM_ (zip changeSeqs priceSeqs) $ \(changeSeq, priceSeq) -> do
        seen :: STUArray s Change Bool
            <- newArray bounds False
        forM_ (zip changeSeq priceSeq) $ \(change, price) -> do
            alreadySeen <- readArray seen change
            if alreadySeen then return ()
            else do
                writeArray seen change True
                oldVal <- readArray totals change
                writeArray totals change (oldVal + price)
    return totals

File: src/22.py
#!/usr/bin/env python

import sys
from itertools import islice
from typing import Iterator
from functools import reduce
from collections import defaultdict

def main():
    input = map(int, open(sys.argv[1]).read().strip().splitlines())
    secret_seqs = map(evolve_secret, input)
    price_seqs = map(lambda seq: map(price, seq), secret_seqs)
    change_seqs = map(lambda seq: (sequence4s(changes(seq))), price_seqs)
    # seq_to_price = [{
    #     seq: price
    #     for (seq, price) in zip(islice(seqs, 1997), islice(prices, 4, None))
    # } for (seqs, prices) in zip(change_seqs, price_seqs)]
    # all_keys = reduce(lambda x, y: x | y.keys(), seq_to_price, set())
    # totals = {
    #     k: sum([m.get(k, 0) for m in seq_to_price])
    #     for k in all_keys
    # }
    totals = compute_totals(change_seqs, price_seqs)
    print(max(totals.values()))

def evolve_secret(num):
    while True:
        num = step(num)
        yield num

def step(x):
    x = prune(mix(x, x << 6))
    x = prune(mix(x, x >> 5))
    x = prune(mix(x, x << 11))
    return x

def prune(x):
    return x & 16777215

def mix(x, y):
    return x ^ y

def price(x):
    return x % 10

def changes(seq: Iterator[int]):
    return (b - a for a, b in zip(seq, islice(seq, 1, None)))

def sequence4s(seq):
    return (
        (a+10)*8000 + (b+10)*400 + (c+10)*20 + d
        for a, b, c, d in
        zip(seq, islice(seq, 1, None), islice(seq, 2, None), islice(seq, 3, None))
    )

def compute_totals(change_seqs, price_seqs):
    totals = defaultdict(lambda: 0)
    for changes, prices in zip(change_seqs, price_seqs):
        seen_seqs = set()
        for change, price in zip (islice(changes, 1997), islice(prices, 4, None)):
            if change in seen_seqs:
                continue
            seen_seqs.add(change)
            totals[change] += price
    return totals

if __name__ == '__main__':
    main()
